crates/python/src/py_sync.rs

```diff
@@ -113,14 +113,16 @@ macro_rules! py_function_sync_async {
         #[allow(clippy::too_many_arguments)]
         #[pyo3(name = $name "")]
         pub fn [< py_ $name >](py: ::pyo3::Python<'_> $(, $(#[$arg_meta])*$arg: $kind)*) $(-> $ret)? {
-            $crate::py_sync::py_sync!(py, $name($($arg),*))
+            use opentelemetry::trace::FutureExt;
+            $crate::py_sync::py_sync!(py, $name($($arg),*).with_current_context())
         }
 
         $(#[$meta])+
         #[pyo3(name = $name "_async")]
         #[allow(clippy::too_many_arguments)]
         pub fn [< py_ $name _async >](py: ::pyo3::Python<'_> $(, $(#[$arg_meta])*$arg: $kind)*) -> ::pyo3::PyResult<&::pyo3::PyAny> {
-            $crate::py_sync::py_async!(py, $name($($arg),*))
+            use opentelemetry::trace::FutureExt;
+            $crate::py_sync::py_async!(py, $name($($arg),*).with_current_context())
```

Comment:
I don't think this should cause issues, but these macros are sensitive to deadlocking and/or blocking signals, so that functionality is worth some extra care when testing these changes.

Reply:
Is there an example of this behavior I can reference?

`FutureExt::with_context`'s main job is to attach the current context: https://docs.rs/opentelemetry/latest/src/opentelemetry/trace/context.rs.html#445-450

Attaching does use a `RefCell`; https://docs.rs/opentelemetry/latest/src/opentelemetry/context.rs.html#11 and RefCell [panics rather than deadlocks](https://doc.rust-lang.org/std/cell/index.html#refcellt) on an attempted illegal borrow.

Reply:
Testing this is non-obvious, so I've made a note to document it better. In short, we just need to make sure functions wrapped by this code (1) don't deadlock, and (2) don't monopolize the GIL and check for signals.

The way I do this is by testing `qcs_sdk.compiler.quilc.compile_program` (and `compile_program_async`). If things compile as expected, then great! It's not deadlocking. I then repeat the same test, but without the `quilc` server running. The call should hang while waiting to connect to the `quilc` server and you should be able to cancel it with `Ctrl-C`. If you can, then it's not monopolizing the GIL and Python is able to do signal handling on the main thread. 

I went ahead and ran this test and these changes ran just fine! I'll resolve this.